### ISSUE TITLE
using videoparse in test client application by using custom gst 

### ISSUE THREAD

**Issue by suganthikarthick** [EricssonResearch/openwebrtc#437]
Hi,

We are trying to read a raw video source file and use it in the test client application by using gustom gst source. 

We have added custom gst video source in the test client application as follows.
#define CUSTOM_VIDEO_SOURCE "/gst_test_video21 ! videoparse width=640 height=360 format=4 ! videoscale"

The command we used for creating the raw video source file is as follows.
gst-launch-1.0 v4l2src ! video/x-raw, width=640, height=360 ! filesink location=/gst_test_video21

When we try to make a call in the application, we are getting the following error.

0:00:08.402580793 13825       0xf19060 ERROR      owrtransportagent owr_transport_agent.c:777:handle_new_send_source: Failed to link "(null)" with transport bin
==== Error message start ====
Error in element videoparse0.
Error: Internal data stream error.
Debugging info: gstrawparse.c(469): gst_raw_parse_loop (): /GstPipeline:transport-agent-1/GstBin:bin1/GstVideoParse:videoparse0:
stream stopped, reason not-negotiated
==== Error message stop ====

But it is working fine with the test_gst_io.c application. We are able to see the video fine. But
When trying to use the same CUSTOM_VIDEO_SOURCE in test_client application and sending it to the cloud, we are facing some issues.

Please suggest on this issue.

**Comment by suganthikarthick** [comment]
SImilarly, we are trying to use read an mp4 file and use it in our application.

We are using the pipeline as follows.
#define CUSTOM_VIDEO_SOURCE "filesrc location=/home/vchn107/temp4_hls/SampleVideo_640x360_30mb.mp4 ! decodebin

But we are facing issue in using decodebin. Please help us on resolving this.
Thanks.

**Comment by superdump** [comment]
If you want to play from a file, maybe you would like to give https://github.com/EricssonResearch/openwebrtc/pull/436 a try.

**Comment by suganthikarthick** [comment]
Hi,

We added the owr_uri related files based on your commit and the application is compiled.
How to test the MP4 file from the application?

We tried to give a MP4 file as the argument to the test-uri application but we are facing segmentation fault. 
What is the command line argument that we we have to give when running the test-uri application?

Thanks.

**Comment by superdump** [comment]
I haven't done much in the way of testing corner cases for the test-uri utility as it was an initial run at it. It expects a URI string as an argument, e.g. file:///path/to/my/file.mp4

Can you provide a backtrace of the segfault?

**Comment by suganthikarthick** [comment]
vchn107@VCHN107:~/cerbero/sources/linux_x86_64/openwebrtc-0.3.0/tests$ ./test-uri file:///home/mk_ashwin/Downloads/Siricha.mp4
==== Error message start ====
Error in element source.
Error: Resource not found.
Debugging info: gstfilesrc.c(514): gst_file_src_start (): /GstPipeline:uri-source-agent-1/GstURIDecodeBin:uridecodebin-1/GstFileSrc:source:
No such file "/home/mk_ashwin/Downloads/Siricha.mp4"
==== Error message stop ====
==== Error message start ====
Error in element source.
Error: Resource not found.
Debugging info: gstfilesrc.c(514): gst_file_src_start (): /GstPipeline:uri-source-agent-1/GstURIDecodeBin:uridecodebin-1/GstFileSrc:source:
No such file "/home/mk_ashwin/Downloads/Siricha.mp4"
==== Error message stop ====

**Comment by superdump** [comment]
`ls -l /home/mk_ashwin/Downloads/Siricha.mp4`?

**Comment by suganthikarthick** [comment]
-rwxrwxrwx 1 mk_ashwin mk_ashwin 97266536 May 12 00:15 /home/mk_ashwin/Downloads/Siricha.mp4

**Comment by superdump** [comment]
I see that you're trying to play a file from another user's home directory. That should be fine given the permissions, but perhaps it is a problem for some as yet unknown reason. Can you try copying the file to your user and making your user the owner?

**Comment by suganthikarthick** [comment]
:) sorry for the blunder mistake. tried to take ssh in another machine and testing.
It is working now. We are planning to merge it with the test client and use it in cloud. That is possible right?

**Comment by superdump** [comment]
I don't really understand what you mean by "use it in cloud" - can you clarify? You can absolutely merge it with the test client and I would welcome a patch that does that.

**Comment by suganthikarthick** [comment]
What I meant is, from test client applciation, instead of taking the local video sources(webcam), we can send this mp4 source as input and the video can be seen from the browser(demo.openwebrtc.io:38080) right?

**Comment by superdump** [comment]
Yes.

**Comment by suganthikarthick** [comment]
Hi, We are not seeing the video in the browser side.
Can you explain the use of the function "owr_uri_source_agent_play(uri_source_agent)" 

Thanks.

**Comment by superdump** [comment]
The behaviour is such that when you create the OwrURISourceAgent and set the URI, that data source is probed and the different audio / video tracks from that URI are exposed as OwrURISources which can then be connected up to renderers or used as send sources on media sessions.

In order to do the probing, we have to set the OwrURISourceAgent to paused. Then when we are ready for media to start flowing, we can set it to playing. If you follow the test_uri.c example, it should work.

**Comment by suganthikarthick** [comment]
yes. with test_uri example. it is working well. We are trying to send the source by setting the source to media session, but that is not playing the media over the cloud. Thanks.

**Comment by superdump** [comment]
You will need to modify test_client.c a bit to make it use OwrURISourceAgent instead of owr_get_capture_sources(). You could start playing after the answer has been sent.

**Comment by suganthikarthick** [comment]
Yes. We made those changes.

**Comment by suganthikarthick** [comment]
I can send the modified application to you. Thanks.

**Comment by suganthikarthick** [comment]
Hi,
I have mailed the modifed testclient application which uses uri sources (an mp4 file) instead of capturing from webcam. 
Thanks.

**Comment by mkwal** [comment]
Hi suganthikarthick,

If possible, can I have this modified test client source code. I am working on same thing (Using C code, participate in session for audio/video). This test_client source code can help me to understand base.

Regards
Manoj

**Comment by slshung** [comment]
Hi suganthikarthick:
Please send mail to me for the modified test_client.c, I need the function too from mp4 file instead of local webcam source in test_client.c.

Best Regards
SL
